File: Binary_Tree_Paths.cpp
//Given a binary tree, return all root-to-leaf paths.

class Solution {
public:
    vector<string> binaryTreePaths(TreeNode* root) {
        vector<string> ans;
        if (root != NULL) BTP(root, "", ans);
        return ans;
    }
    
    void BTP(TreeNode* root, string path, vector<string>& ans){
        if (root -> left == NULL && root -> right == NULL) ans.push_back(path + to_string(root -> val));
        if (root -> left != NULL) BTP(root -> left, path + to_string(root -> val) + "->",ans);
        if (root -> right != NULL) BTP(root -> right, path + to_string(root -> val) + "->",ans);
    }
};

File: lesson_5_binaryTree/BinaryTreePaths.java
package leetcode.lesson_5_binaryTree;

import java.util.ArrayList;
import java.util.List;

public class BinaryTreePaths {

    public List<String> binaryTreePaths(TreeNode root) {
        List<String> ans = new ArrayList<>();
        if (root.right == null && root.left == null) {
            ans.add("" + root.val);
            return ans;
        }

        List<String> lefts = binaryTreePaths(root.left);
        for (int i = 0; i < lefts.size(); i ++){
            ans.add(root.val + "->" + lefts.get(i));
        }

        List<String> rights = binaryTreePaths(root.right);
        for (int i = 0; i < rights.size(); i ++){
            ans.add(root.val + "->" + rights.get(i));
        }
        return ans;
    }

}
